### PR TITLE
Final PR of group n-te Gruppe

### PR DESCRIPTION
**Pull request by natalie-bier**
## Motivation and Context
Final outcome of the group’s first two weeks of work
---

## Public API Changes

-   [x ] No Public API changes
-   [ ] Yes, Public API changes (Details below)

---

## How Has This Been Tested?

---

## Checklist

-   [x ] The changes have been tested locally.
-   [ ] Documentation has been updated (if the public API or usage changes).
-   [ ] A entry has been added to [`CHANGELOG.md`](https://github.com/pwhofman/probly/blob/main/CHANGELOG.md) (if relevant for users).
-   [ ] The code follows the project's [style guidelines](https://github.com/pwhofman/probly/blob/main/.github/CONTRIBUTING.md) and passes minimal code style checks.
-   [ ] I have considered the impact of these changes on the public API.

---

### CODE REVIEW

File: pytest.ini
Comment (Cortys) [inline pytest.ini]: pytest is already configured in `pyproject.toml`, this ini file is not needed.

File: test_flax_ensemble.py
Comment (Cortys) [inline test_flax_ensemble.py]: Test files used during development should not be part of the PR. Please remove.

File: src/probly/layers/flax.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
+from __future__ import annotations
+
+import jax
+from flax import nnx
+from flax.nnx import rnglib   
+
+
+class DropConnectDense(nnx.Module):
```
Comment (Cortys) [inline src/probly/layers/flax.py]: Missing docstring

File: src/probly/layers/flax.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    
+ def __call__(self, x: jax.Array, *, deterministic: bool = False) -> jax.Array:
+     if not deterministic:
+        key = self.rngs.dropout()
+        mask = jax.random.uniform(key, shape=self.weight.value.shape) > self.p
+        weight = self.weight.value * mask  
+     else:
+        weight = self.weight.value * (1.0 - self.p)
+     y = x @ weight.T
+     if self.bias is not None:
```
Comment (Cortys) [inline src/probly/layers/flax.py]: `base_layer.kernel` is already a `Param`. By calling `Param` on it again, the underlying `Array` is wrapped twice.

Comment (Cortys) [inline src/probly/layers/flax.py]: `__call__` should optionally also accept `rngs`. This allows the used to provide a PRNG either at construction or at call time.
Providing the PRNG at call time is the preferred way of doing this in flax as it allows for more explicit control over stochasticity (see [Randomness in flax](https://flax.readthedocs.io/en/latest/guides/randomness.html)).

Take a look at the implementation of `Dropout` in flax. The implementation of `DropConnectDense` should roughly follow it.

```suggestion
    def __call__(
        self,
        x: jax.Array,
        *,
        deterministic: bool = False,
        rngs: rnglib.Rngs | rnglib.RngStream | jax.Array | None = None,
    ) -> jax.Array:
```

File: src/probly/transformation/dropconnect/flax.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+import flax.nnx as nnx
+
+from probly.layers.flax import DropConnectDense
+
+from .common import register
+
+
+def replace_flax_dropconnect(obj: nnx.Linear, p: float) -> DropConnectDense:
+    """Replace a given layer by a DropConnectDense layer."""
+    rngs = nnx.Rngs(0)
```
Comment (Cortys) [inline src/probly/transformation/dropconnect/flax.py]: The PRNG should not be reinitialized for each DropConnect layer, this makes the model less random and also removes control over the randomness from the downstream user.
To address this, the PRNG should ideally be passed in explicitly (see comments on the flax `DropConnect` layer) (see #191).

For now, please simply remove the `rngs` here. A more general approach can then be added in a future PR.

File: tests/probly/transformation/dropconnect/test_common.py
Comment (Cortys) [inline tests/probly/transformation/dropconnect/test_common.py]: Common tests should test any `torch` specifics. Only the general `dropconnect` API should be tested here.

File: tests/probly/transformation/dropconnect/_init_.py
Comment (Cortys) [inline tests/probly/transformation/dropconnect/_init_.py]: Add a docstring.

File: tests/probly/transformation/evidential/regression/test_torch.py
Comment (Cortys) [inline tests/probly/transformation/evidential/regression/test_torch.py]: See the commments regarding `test_common` of evidential regression.
The torch tests should check whether the *external* bevahior of the `evidential_regression` transformation is correct, i.e., whether it correctly inserts layers at the ent of the module.

More specifically, I suggest that you use some of the torch fixtures that are, for example, already used in the `Dropout` tests, simply apply the `evidential_regression` transformation on it, and then inspect the transformed model.
No importlib hacks (or similar) are required here.

File: notebooks/examples/dropout_transformation.ipynb
Comment (Cortys) [inline notebooks/examples/dropout_transformation.ipynb]: Nice introduction to Dropout! Comments:
1. In part B.2, the code cell does not work (`enable_at_eval` is not a valid param). Please fix this.
2. There are Ruff warnings for some of the code, please address them.

**Review by Cortys — CHANGES_REQUESTED** [review]
Thank you for this PR! Overall, you did a great job integrating the requested changes.

Please make sure that your PR passes **all** CI steps, i.e., all tests should pass and there should be no Ruff or mypy errors.
Please also carefully read my detailed comments on each file and try to address the raised points.

File: .github/workflows/ci.yml
[the remark this answers is not in the document]
Comment (Serurays) [reply .github/workflows/ci.yml]: Done!

File: notebooks/examples/dropout_transformation.ipynb
Comment (Serurays) [reply notebooks/examples/dropout_transformation.ipynb]: Done!

File: src/probly/layers/flax.py (hunk 1 above)
Comment (Serurays) [reply src/probly/layers/flax.py]: Done!

File: tests/probly/transformation/dropconnect/test_common.py
Comment (Serurays) [reply tests/probly/transformation/dropconnect/test_common.py]: Done!

File: tests/probly/transformation/evidential/regression/test_torch.py
Comment (Serurays) [reply tests/probly/transformation/evidential/regression/test_torch.py]: Done!

File: .github/workflows/ci.yml
[the remark this answers is not in the document]
Comment (Cortys) [reply .github/workflows/ci.yml]: Please undo the remaining change (set `fail_ci_if_error: true`) so that the workflows for your PR can run.
